orttraining/orttraining/core/graph/graph_parser/pattern_graph.h

```diff
@@ -0,0 +1,299 @@
+
+  NodeDef GetNodeDef() {
+    auto IA = [](const std::string& argSuffix, const TypeProto* type_proto = nullptr) {
+      return ArgDef(argSuffix, type_proto);
+    };
+    std::string domain = domains.empty() ? "" : domains[0];
+    int version = versions.empty() ? 9 : versions[0];
+    std::vector<ArgDef> input_args(input_args_name.size());
+    std::vector<ArgDef> output_args(output_args_name.size());
+    // give a name if the name is not given by user
+    for (size_t i = 0; i < input_args_name.size(); i++) {
+      input_args[i] = IA(input_args_name[i]);
+    }
+    for (size_t i = 0; i < output_args_name.size(); i++) {
+      output_args[i] = IA(output_args_name[i]);
+    }
+    return NodeDef(OpDef(op_type_, domain, version), input_args, output_args, attributes, GetNodeName());
+  }
+
+ private:
+  std::string op_type_;
+  std::vector<std::string>& input_args_name;
+  std::vector<std::string>& output_args_name;
+  std::string node_name_;
+  std::vector<std::string>& domains;
+  std::vector<int>& versions;
+  std::vector<AttributeProto>& attributes;
+  std::unordered_set<std::string> valid_domains;
+  std::unordered_set<int> valid_versions;
+};
+
+struct PatternGraph;
+
+class NodeCompareFunc {
+ public:
+  virtual bool operator()(const Node* g, const PNode* p, const Graph& /*target*/, const PatternGraph& /*pattern*/) const = 0;
+};
+
+class DefaultNodeCompareFunc : public NodeCompareFunc {
+ public:
+  DefaultNodeCompareFunc(bool skip_optype, bool skip_domain,
+                         bool skip_version, bool skip_path)
+      : skip_optype_(skip_optype),
+        skip_domain_(skip_domain),
+        skip_version_(skip_version),
+        skip_path_(skip_path) {}
+
+  bool operator()(const Node* g, const PNode* p, const Graph& /*target*/, const PatternGraph& /*pattern*/) const override {
+    if (!g && !p)
+      return true;
+    if (!g && p || !p && g)
+      return false;
+    if (!skip_optype_ && p->NameEquals(g->OpType())) {
+      // PARSER_LOG << "OpType mismatch, "
+      //            << "g is: " << g->OpType() << ", p is: " << p->OpType();
+      return false;
+    }
+    if (!skip_domain_ && p->DomainsContain(g->Domain())) {
+      // PARSER_LOG << "Domain mismatch, "
+      //            << "g is: " << g->Domain() << ", p is: " << p->Domain();
+      return false;
+    }
+    if (!skip_version_ && p->VersionsContain(g->SinceVersion())) {
+      // PARSER_LOG << "Version mismatch, "
+      //            << "g is: " << g->SinceVersion() << ", p is: " << p->SinceVersion();
+      return false;
+    }
+    return true;
+  }
+
+  bool skip_optype_, skip_domain_, skip_version_, skip_path_;
+};
+
+// typedef std::function<bool(const Node*, const PNode*, const Graph&, const PatternGraph&)> NodeCompareFunc;
+
+struct PatternGraph {
+  PatternGraph(const std::vector<NodeDef>& constants, const std::vector<PNode>& pnodes,
+               const std::string& root_node = "", const std::string& name = "")
```

Comment:
>root_node [](http://example.com/codeflow?start=34&length=9)

do we need user to pass this？ I mean， can we get that by parsing the NodeDef?

Reply:
What about move this parameter into ```TryMatch```? We may add an optional parameter named ```seachStartNodeName``` to tell user he can specify it. I think in this way it could be more clear.